// hdd-space/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var child_process_1 = require("child_process");
var utils_1 = require("./utils");
var format_1 = require("./format");
var unix_1 = require("./unix");
var win32_1 = require("./win32");
function formatResult(opts, res) {
    var formater = function (size) { return format_1.formatSize(opts.format, size); };
    var newRes = {
        parts: res.parts.map(function (part) {
            var newPart = {
                free: formater(part.free),
                size: formater(part.size),
                place: part.place
            };
            if ('letter' in part) {
                newPart.letter = part.letter;
            }
            if ('mountOn' in part) {
                newPart.mountOn = part.mountOn;
            }
            return newPart;
        }),
        total: {
            free: formater(res.total.free),
            size: formater(res.total.size)
        }
    };
    return newRes;
}
function runCmd(commandFn, parser, callback) {
    commandFn(function (err, output) {
        if (err) {
            callback(err);
            return;
        }
        if (!output || output.trim() === '') {
            callback(new Error('empty output'));
            return;
        }
        try {
            var result = parser(output);
            if (result.parts.length === 0) {
                throw new Error('cannot parse output');
            }
            callback(null, result);
        }
        catch (e) {
            callback(new Error('cannot parse output'));
        }
    });
}
var defaultOpts = {
    format: function (size) { return size; }
};
function getHddInfo(opts, callback) {
    var platform = opts.platform || process.platform;
    var _a = platform === 'win32'
        ? [win32_1.win32Cmd, win32_1.parseWin32Output]
        : [unix_1.unixCmd, unix_1.parseUnixOutput], cmd = _a[0], parser = _a[1];
    var fn = opts.fetchOutput || ('output' in opts
        ? function (cb) { return cb(null, opts.output); }
        : child_process_1.exec.bind(null, cmd));
    runCmd(fn, parser, function (err, result) {
        if (err) {
            callback(err);
            return;
        }
        var formatted = formatResult(opts, result);
        callback(null, formatted);
    });
}
exports.getHddInfo = getHddInfo;
function fetchHddInfo(opts) {
    if (opts === void 0) { opts = defaultOpts; }
    return new Promise(function (resolve, reject) {
        getHddInfo(opts, function (err, res) {
            if (err) {
                reject(err);
                return;
            }
            resolve(res);
        });
    });
}
exports.fetchHddInfo = fetchHddInfo;
function getCrossPlatformInfo(opts, callback) {
    if (arguments.length < 2) {
        callback = opts;
        opts = defaultOpts;
    }
    opts = opts;
    getHddInfo(opts, function (err, res) {
        if (err) {
            throw err;
        }
        callback(res);
    });
}
exports.default = getCrossPlatformInfo;
module.exports = utils_1.assign(getCrossPlatformInfo, module.exports);
