// hdd-space/dist/unix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var parse_columns_1 = require("parse-columns");
function parseDf(output) {
    var extractTable = {
        3: 'free',
        5: 'mp',
        1: 'size'
    };
    var res = [];
    parse_columns_1.default(output, {
        transform: function (value, header, columnIndex, rowIndex) {
            var key = extractTable[columnIndex];
            if (!key) {
                return;
            }
            var row = res[rowIndex - 1];
            if (!row) {
                row = {};
                res[rowIndex - 1] = row;
            }
            row[key] = key === 'mp' ? value : Number(value);
        }
    });
    return res;
}
/*
        [df] output example:

        Filesystem     1K-blocks    Used Available Use% Mounted on
        rootfs          15506408 1149728  13708188   8% /
        /dev/root       15506408 1149728  13708188   8% /
        devtmpfs          447624       0    447624   0% /dev
        tmpfs              89548     208     89340   1% /run
        tmpfs               5120       0      5120   0% /run/lock
        tmpfs             179080       0    179080   0% /run/shm

    */
function parseUnixOutput(output) {
    var parsed = parseDf(output.trim());
    var root = null;
    var parts = parsed
        .map(function (part) {
        var res = {
            size: part.size * 1024,
            free: part.free * 1024,
            place: part.mp,
            mountOn: part.mp
        };
        if (part.mp === '/') {
            root = res;
        }
        return res;
    });
    var total = {
        size: root.size,
        free: root.free
    };
    return {
        parts: parts,
        total: total
    };
}
exports.parseUnixOutput = parseUnixOutput;
exports.unixCmd = 'df -Pk';
