// hdd-space/dist/win32.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getTotal(parts) {
    return parts.reduce(function (total, part) {
        total.size += part.size;
        total.free += part.free;
        return total;
    }, {
        size: 0,
        free: 0
    });
}
/*
    [wmic logicaldisk get size,freespace,caption] output example:

    Caption  FreeSpace	Size
    C:	   1286164480   34359734272
    D:	   1864638464   50925137920
    E:
    F:	   77553082368  990202818560
    G:
    L:
*/
function parseWin32Output(output) {
    var parts = output
        .split('\n')
        .slice(1) // remove header
        .map(function (line) {
        var lineParts = line.split(/[\s]+/g);
        var partInfo = {
            place: lineParts[0],
            letter: lineParts[0],
            free: parseInt(lineParts[1]),
            size: parseInt(lineParts[2])
        };
        if (isNaN(partInfo.free) ||
            isNaN(partInfo.size) ||
            partInfo.place === '') {
            return null;
        }
        return partInfo;
    })
        .filter(function (part) {
        return !!part;
    });
    return {
        parts: parts,
        total: getTotal(parts)
    };
}
exports.parseWin32Output = parseWin32Output;
exports.win32Cmd = 'wmic logicaldisk get size,freespace,caption';
